src/hyperlight_host/src/hypervisor/mod.rs

```diff
@@ -159,26 +167,264 @@ pub(crate) trait Hypervisor: Debug + Send {
     fn dispatch_call_from_host(
         &mut self,
         dispatch_func_addr: RawPtr,
+        mem_mgr: &mut SandboxMemoryManager<HostSharedMemory>,
+        host_funcs: Arc<Mutex<FunctionRegistry>>,
         #[cfg(gdb)] dbg_mem_access_fn: Arc<Mutex<SandboxMemoryManager<HostSharedMemory>>>,
     ) -> Result<()>;
 
-    /// Handle an IO exit from the internally stored vCPU.
-    fn handle_io(
-        &mut self,
-        port: u16,
-        data: Vec<u8>,
-        rip: u64,
-        instruction_length: u64,
-    ) -> Result<()>;
+    // Run the vCPU "one step until it exits"
+    fn run_vcpu(&mut self) -> Result<VmExit>;
 
-    /// Run the vCPU
+    /// Run the vCPU while handling potential exists. Returns when the vCPU halts or an error occurs.
+    #[allow(clippy::too_many_arguments)]
     fn run(
         &mut self,
-        #[cfg(feature = "trace_guest")] tc: &mut crate::sandbox::trace::TraceContext,
-    ) -> Result<HyperlightExit>;
+        _entrypoint: u64,
+        interrupt_handle: Arc<dyn InterruptHandleImpl>,
+        sandbox_regions: &[MemoryRegion],
+        mmap_regions: &[MemoryRegion],
+        mem_mgr: &mut SandboxMemoryManager<HostSharedMemory>,
+        host_funcs: Arc<Mutex<FunctionRegistry>>,
+        #[cfg(gdb)] dbg_mem_access_fn: Arc<Mutex<SandboxMemoryManager<HostSharedMemory>>>,
+        #[cfg(crashdump)] rt_cfg: &SandboxRuntimeConfig,
+    ) -> Result<()> {
+        // ===== KILL() TIMING POINT 1: Between guest function calls =====
+        // Clear any stale cancellation from a previous guest function call or if kill() was called too early.
+        // This ensures that kill() called BETWEEN different guest function calls doesn't affect the next call.
+        //
+        // If kill() was called and ran to completion BEFORE this line executes:
+        //    - kill() has NO effect on this guest function call because CANCEL_BIT is cleared here.
+        //    - NOTE: stale signals can still be delivered, but they will be ignored.
+        interrupt_handle.clear_cancel();
+
+        // Keeps the trace context and open spans
+        #[cfg(feature = "trace_guest")]
+        let mut tc = crate::sandbox::trace::TraceContext::new();
+
+        let result = loop {
+            // ===== KILL() TIMING POINT 2: Before set_tid() =====
+            // If kill() is called and ran to completion BEFORE this line executes:
+            //    - CANCEL_BIT will be set and we will return an early VmExit::Cancelled()
+            interrupt_handle.set_tid();
+            interrupt_handle.set_running();
+
+            let exit_reason =
+                if interrupt_handle.is_cancelled() || interrupt_handle.is_debug_interrupted() {
+                    Ok(VmExit::Cancelled())
+                } else {
+                    #[cfg(feature = "trace_guest")]
+                    tc.setup_guest_trace(tracing_opentelemetry::OpenTelemetrySpanExt::context(
```

Comment:
Even though this looks a lot cleaner, the `setup_guest_trace` is supposed to be run right before entering the `guest` because it takes a reading from the TSC so that it can correctly calculate the time spent in the guest.
Additionally, it stores the current span context as a parent for all subsequent guest spans. In this case, there is an additional span created by the call to `self.run_vcpu` which should be the parent of all the guest spans, but it will not be, as the current context is set as the parent.

Reply:
I've been looking for an alternative to this when I was working on tracing, but getting the context for a function can only be done with that `tracing::Span::current()`, which yields the Span for the current function (the last span created).

Reply:
Oopps, good catch. Do you think it's sufficient to ensure `run_vcpu` does not emit a span (and i guess is part of caller `run`?